src/Libraries/Microsoft.Extensions.Telemetry.Testing/Metering/CollectedMeasurement.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.Telemetry.Testing.Metering;
+
+/// <summary>
+/// Represents a single measurement performed by an instrument.
+/// </summary>
+/// <typeparam name="T">The type of metric measurement value.</typeparam>
+[Experimental]
+public sealed class CollectedMeasurement<T>
+    where T : struct
+{
+    /// <summary>
+    /// Initializes a new instance of the <see cref="CollectedMeasurement{T}"/> class.
+    /// </summary>
+    /// <param name="value">The measurement's value.</param>
+    /// <param name="tags">The dimensions of this measurement.</param>
+    /// <param name="timestamp">The time that the measurement occurred at.</param>
+    internal CollectedMeasurement(T value, ReadOnlySpan<KeyValuePair<string, object?>> tags, DateTimeOffset timestamp)
+    {
+        var d = new Dictionary<string, object?>();
+        foreach (var tag in tags)
+        {
+            d[tag.Key] = tag.Value;
```

Comment:
This will lose duplicate tag values. Last tag with a name wins. Do we care?

Reply:
I asked Noah about this yesterday, and there doesn't seem to be a consensus on first-one-wins or last-one-wins. This is definitely interpreted as a dictionary at the end of the day. So I opted for last-one-wins.

Reply:
> This is definitely interpreted as a dictionary at the end of the day

Some scenarios treat it as a list and don't remove duplicates at all :) I have no qualms with this though.